File: backend/h3_timing.py
"""Frame-based H3 generation snapshots; independent of mutable editor settings."""
import re


H3_SUFFIX = re.compile(r"__h3v([12])_c(\d+)_r(\d+)_h(\d+)_t(\d+)_f(\d+)_s([01])(?:_n(\d+))?(?=\.[^.]+$)")


def timing_from_filename(path):
    match = H3_SUFFIX.search(str(path))
    if not match:
        return None
    version, context, raw, head, tail, fps_milli, save = map(int, match.groups()[:7])
    carry = int(match.group(8) or 0)
    if (version == 2 and match.group(8) is None) or (version == 1 and carry):
        raise ValueError("Invalid H3 context carry in video filename.")
    if fps_milli <= 0 or not 0 <= carry <= context or raw <= context - carry + head + tail:
        raise ValueError("Invalid H3 timing in video filename.")
    return dict(version=version, context_frames=context, raw_frames=raw, context_carry_frames=carry,
                head_frames=head, tail_frames=tail, fps=fps_milli / 1000,
                play_frames=raw - context + carry - head - tail, save_latent=bool(save))


def timing_filename(path, timing):
    stem, dot, ext = str(path).rpartition(".")
    if not dot:
        return path
    stem = H3_SUFFIX.sub("", str(path)).rpartition(".")[0]
    version = timing.get("version", 1)
    carry = f"_n{timing.get('context_carry_frames', 0)}" if version == 2 else ""
    return (f"{stem}__h3v{version}_c{timing['context_frames']}_r{timing['raw_frames']}"
            f"_h{timing['head_frames']}_t{timing['tail_frames']}"
            f"_f{round(timing['fps'] * 1000)}_s{int(timing['save_latent'])}{carry}.{ext}")


def choose_h3_context(previous, requested=22):
    """Keep raw_end - context inside the previous visible source interval."""
    visible_start = previous["context_frames"] - previous.get("context_carry_frames", 0) + previous["head_frames"]
    low = previous["tail_frames"] + 1
    high = previous["raw_frames"] - visible_start
    first = 5 + max(0, (low - 5 + 16) // 17) * 17
    candidates = range(first, high + 1, 17)
    if not candidates:
        raise ValueError("H3 context cannot start inside the previous clip's visible frames; increase its duration or reduce its head/tail extensions.")
    return min(candidates, key=lambda value: (abs(value - requested), value))


def plan_h3_clips(clips, fps):
    previous_by_track = {}
    plans = []
    for clip in clips:
        track = clip.get("z_index", 0)
        previous = previous_by_track.get(track)
        start = clip["preview_start_ms"]
        end = clip["preview_end_ms"]
        is_h3 = clip.get("agent", "MiniMaxH3") == "MiniMaxH3"
        linked = bool(is_h3 and previous and previous.get("h3_timing")
                      and previous.get("save_latent") and previous.get("z_index", 0) == track
                      and abs(previous["preview_end_ms"] - start) <= 1)
        requested = int(clip.get("h3_motion_context_length", 0)) or (22 if linked else 0)
        if is_h3 and (requested or clip.get("save_latent")):
            preferred = max(5, (requested - 5) // 17 * 17 + 5)
            context = choose_h3_context(previous["h3_timing"], preferred) if linked else 0
            if linked:
                clip["h3_motion_context_length"] = context
            carry = previous["h3_timing"]["tail_frames"] if linked else 0
            head = max(0, round((start - clip["start_ms"]) * fps / 1000))
            tail = max(0, round((clip["end_ms"] - end) * fps / 1000))
            target = round(end * fps / 1000) - round(start * fps / 1000)
            required = max(5, target + context - carry + head + tail)
            raw = required + (5 - required) % 17
            tail = raw - (context - carry + head + target)
            play = target
            if play <= 0:
                raise ValueError("H3 trim removes the entire clip.")
            play_start = round(start * fps / 1000)
            clip["h3_timing"] = dict(
                version=2, fps=fps, requested_context_frames=requested, context_carry_frames=carry,
                context_frames=context, raw_frames=raw, head_frames=head,
                tail_frames=tail, play_frames=play, save_latent=bool(clip.get("save_latent")),
                play_start_frame=play_start, play_end_frame=play_start + play,
                source_start_ms=start, source_end_ms=end,
                previous_source_clip_id=previous.get("source_clip_id") if context else None,
            )
            if clip.get("output_video"):
                clip["output_video"] = timing_filename(clip["output_video"], clip["h3_timing"])
            plans.append(clip)
        previous_by_track[track] = clip

    for clip in plans:
        timing = clip["h3_timing"]
        if clip.get("output_video"):
            clip["output_video"] = timing_filename(clip["output_video"], timing)
        clip["playback_spans"] = [dict(source_clip_id=clip.get("source_clip_id"),
            output_video=clip.get("output_video"), start_frame=timing["context_frames"] - timing["context_carry_frames"] + timing["head_frames"],
            frame_count=timing["play_frames"])]
    by_id = {c.get("source_clip_id"): c for c in plans}
    for clip in plans:
        timing = clip["h3_timing"]
        previous = by_id.get(timing["previous_source_clip_id"])
        if not previous:
            continue
        prior = previous["h3_timing"]
        # Context includes the preceding raw video's padding, which is not visible.
        overlap = timing["context_frames"] - prior["tail_frames"]
        previous["playback_spans"][0]["frame_count"] -= overlap
        previous["playback_spans"].append(dict(source_clip_id=clip.get("source_clip_id"),
            output_video=clip.get("output_video"), start_frame=0,
            frame_count=overlap, role="regenerated_context"))


def source_clip_timing(clip):
    """Undo only our automatic playback layout, never a user's later resize/move."""
    layout = clip.get("h3_layout")
    if not isinstance(layout, dict) or not all(isinstance(layout.get(key), (int, float)) for key in (
        "start_sec", "duration_sec", "source_start_sec", "source_duration_sec",
    )):
        return clip
    start = float(clip.get("start_ms", 0))
    duration = float(clip.get("duration_ms", float(clip.get("end_ms", start)) - start))
    if abs(start - layout["start_sec"] * 1000) <= 2:
        start = round(layout["source_start_sec"] * 1000)
    if abs(duration - layout["duration_sec"] * 1000) <= 2:
        duration = round(layout["source_duration_sec"] * 1000)
    out = dict(clip, start_ms=start, duration_ms=duration, end_ms=start + duration,
               resource_start_sec=start / 1000, resource_duration_sec=duration / 1000)
    out.pop("h3_layout", None)
    return out


def trim_h3_video(timing, actual_frames, fps):
    if abs(fps - timing["fps"]) > 0.01:
        raise ValueError("H3 video fps differs from the generation snapshot.")
    raw, context = timing["raw_frames"], timing["context_frames"]
    head, tail = timing["head_frames"], timing["tail_frames"]
    carry = timing.get("context_carry_frames", 0)
    tolerance = 0 if timing.get("version", 1) == 2 else 1
    if abs(actual_frames - raw) <= tolerance:
        offset = context - carry + head
    elif abs(actual_frames - timing["play_frames"]) <= tolerance:
        return 0, actual_frames / fps
    elif abs(actual_frames - (raw - context)) <= tolerance:
        if carry:
            raise ValueError("H3 video has already lost context carry frames; use the original untrimmed generation.")
        offset = head
    else:
        raise ValueError(f"H3 video has {actual_frames} frames; generation snapshot expects {raw} raw, {raw - context} context-trimmed or {timing['play_frames']} final frames.")
    return offset / fps, (actual_frames - offset - tail) / fps

File: js/editor/H3Timing.js
const SUFFIX = /__h3v([12])_c(\d+)_r(\d+)_h(\d+)_t(\d+)_f(\d+)_s([01])(?:_n(\d+))?(?=\.[^.]+$)/;

export function stripH3Timing(file) {
    return String(file).replace(SUFFIX, "");
}

export function h3TimingFromFilename(file) {
    const match = String(file).match(SUFFIX);
    if (!match) return null;
    const [version, context, raw, head, tail, fpsMilli, save] = match.slice(1, 8).map(Number);
    const carry = Number(match[8] || 0);
    if ((version === 2 && match[8] == null) || (version === 1 && carry)) return null;
    if (!(fpsMilli > 0 && carry <= context && raw > context - carry + head + tail)) return null;
    return { version, context, carry, raw, head, tail, fps: fpsMilli / 1000, save: !!save, play: raw - context + carry - head - tail };
}

export function applyH3VideoTrim(gen) {
    const timing = h3TimingFromFilename(gen.file);
    if (!timing || gen.h3_trim_applied) return false;
    const actual = Math.round(Number(gen.duration_sec) * timing.fps);
    if (!(actual > 0)) return false;
    let offset;
    let tail = timing.tail;
    const tolerance = timing.version === 2 ? 0 : 1;
    if (Math.abs(actual - timing.raw) <= tolerance) offset = timing.context - timing.carry + timing.head;
    else if (Math.abs(actual - timing.play) <= tolerance) { offset = 0; tail = 0; }
    else if (Math.abs(actual - (timing.raw - timing.context)) <= tolerance) {
        if (timing.carry) throw new Error('H3 context carry frames were removed; use the original untrimmed video.');
        offset = timing.head;
    }
    else throw new Error(`H3 video length (${actual} frames) does not match its generation timing (${timing.raw}/${timing.raw - timing.context}/${timing.play}).`);
    gen.trim_in_sec = offset / timing.fps;
    gen.trim_out_sec = (actual - tail) / timing.fps;
    gen.h3_trim_applied = true;
    return true;
}

// Replace the preceding visible tail with the regenerated context from its successor.
export function replaceH3ContextTail(previousRows, nextRows, previousDuration) {
    const rows = previousRows.filter(row => !row.h3_context_from).map(row => ({ ...row }));
    for (const row of rows) {
        if (row.h3_context_original_out != null) {
            row.trim_out_sec = row.h3_context_original_out;
            delete row.h3_context_original_out;
        }
    }
    const previous = rows.find(row => row.enabled !== false);
    const next = nextRows.find(row => row.enabled !== false && !row.h3_context_from);
    const a = previous && h3TimingFromFilename(previous.file);
    const b = next && h3TimingFromFilename(next.file);
    if (!a?.save || !b?.context || a.fps !== b.fps || !previous.h3_trim_applied || !next.h3_trim_applied) return rows;
    if (b.version === 2 && (a.version !== 2 || b.carry !== a.tail)) return rows;
    if (Math.round(next.duration_sec * b.fps) !== b.raw || Math.round(previous.duration_sec * a.fps) !== a.raw) return rows;
    const end = Math.min(previous.trim_out_sec, previous.trim_in_sec + previousDuration - (previous.edit_start_sec || 0));
    const overlap = Math.min(end - previous.trim_in_sec, Math.max(0, end - (a.raw - b.context) / a.fps));
    if (!(overlap > 0)) return rows;
    previous.h3_context_original_out = previous.trim_out_sec;
    previous.trim_out_sec = end - overlap;
    const prefix = Math.max(0, previous.trim_out_sec - (a.raw - b.context) / a.fps);
    const contextId = `context_${previous.id}_${next.id}`;
    const existing = previousRows.find(row => row.id === contextId);
    rows.unshift({ ...next, id: contextId, h3_context_from: next.id,
        enabled: (existing || next).enabled !== false, muted: (existing || next).muted === true,
        trim_in_sec: prefix, trim_out_sec: prefix + overlap,
        edit_start_sec: (previous.edit_start_sec || 0) + previous.trim_out_sec - previous.trim_in_sec });
    return rows;
}

// One-time repair for projects saved by the removed automatic ripple feature.
export function restoreH3ClipTiming(clip) {
    if (!clip.h3_layout) return clip;
    const out = { ...clip };
    const old = out.h3_layout;
    delete out.h3_layout;
    if (!['start_sec', 'duration_sec', 'source_start_sec', 'source_duration_sec']
        .every(key => Number.isFinite(old[key]))) return out;
    let start = Number(clip.start_ms) || 0;
    let duration = Number(clip.duration_ms) || Number(clip.end_ms) - start;
    if (Math.abs(start - old.start_sec * 1000) <= 2) start = Math.round(old.source_start_sec * 1000);
    if (Math.abs(duration - old.duration_sec * 1000) <= 2) duration = Math.round(old.source_duration_sec * 1000);
    out.start_ms = start;
    out.duration_ms = duration;
    out.end_ms = start + duration;
    out.resource_start_sec = start / 1000;
    out.resource_duration_sec = duration / 1000;
    return out;
}
